### PR TITLE
#97 - New initializer code and general improvements to the architecture

### PR DESCRIPTION
**Pull request by leMicin**
# Warning! There may be bugs or broken features from this PR. I am sorry if I broke anything.

# New initializer
The new initializer allows 3 different hooks in the initialization pipeline. In order they are BeforeInit, Init, AfterInit. Data initialization will mostly be done in one of the first two hooks. The third hook is mostly used to give feedback to the user that the tool is ready.

This allows to clean some of the dependencies. For exemple, the TradeClient was the service that initiated the Tray message saying the tool is ready. It really didn't make sense as the TradeClient should only serve to get the trade data.

To use the new initializer, simply add one of these interfaces to your services : IOnBeforeInit, IOnInit, IOnAfterInit, IOnReset.

# Refactor
In the past, I simply tried to move the code into the new projects and make it work. I now started refactoring the code that was previously moved. One of the exemples of this is the LeagueService. That code was previously inside the TradeClient. For contributers, the LeagueService is the exemple of what the new architecture code aims to achieve in terms of readability, maintainability.

### CODE REVIEW

**Comment by leMicin** [comment]
Referencing this PR inside the issue #97

**Comment by Blinke** [comment]
Good work on the initializer! I think it's going to be a great way to handle things going forward.

Getting the types that implement the different initialization interfaces seems more complex than it needs to be. Right now it scans through all our assemblies (including `mscorlib`) four times to find the appropriate types.

I think we could make the process more streamlined and handle it at the service registration stage by introducing an extensions method like this: 

    private static readonly HashSet<Type> intializeTypes = new HashSet<Type>
    {
        typeof(IOnReset),
        typeof(IOnBeforeInit),
        typeof(IOnInit),
        typeof(IOnAfterInit)
    };
    
    public static void AddInitalizableService<TInterface, TImplementation>(this IServiceCollection services)
        where TInterface : class
        where TImplementation : class, TInterface
    {
        // Hardcoded to singleton because I think that's the only scope that makes sense for an intializable service
        services.AddSingleton<TInterface, TImplementation>();
    
        foreach (var initializeInterface in typeof(TImplementation).GetInterfaces().Where(i => intializeTypes.Contains(i)))
        {
            services.AddSingleton(initializeInterface.GetType(), typeof(TImplementation));
        }
    }

We could then simple inject `IEnumerable<IOnReset>, IEnumerable<OnbeforeInit>`, etc into the Initializer and not have to worry about scanning the assemblies at all.

This would make it clearer at a glance which registered services have initialization logic. We also would only use reflection to scan a few select types for the matching interfaces rather than multiple assemblies.

**Comment by leMicin** [comment]
@Blinke 
I think your way wouldn't work. I think that when the DI tries to get the service for the "real" implementation would have a different instance than when you initialize with the other interfaces.
Exemple:
``` cs
services.AddSingleton(typeof(IServiceA), typeof(ServiceA));
services.AddSingleton(typeof(IOnAfterInit), typeof(ServiceA)); // This wouldn't know to use the previously existing IServiceA implementation
```

Attributes and Reflection are tools that are available to us in the .net ecosystem. I like to remove configuration steps from the developper as much as possible, I agreed to remove the service registration attribute helpers because it wasn't standard practice. The initializer is pretty specific to Sidekick, so I am happy to leave the code as it is this time.

Now you also mentioned performance, and there I agree with you. I changed the code of the FindType to use this line:
``` cs
foreach (var assemblyName in assemblies.Where(x => x.FullName.StartsWith("Sidekick")))
```
instead of the old
``` cs
foreach (var assemblyName in assemblies.Where(x => !x.FullName.StartsWith("Microsoft") && !x.FullName.StartsWith("System")))
```
We are now scanning our own projects only. It is now a lot faster than it was.

File: src/Sidekick/Services/TrayService.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        public TrayService(ILeagueService leagueService)
+        {
+            this.leagueService = leagueService;
+        }
+
+        public Task OnAfterInit()
+        {
+            TrayIcon.ReloadUI();
+            TrayIcon.PopulateLeagueSelectMenu(leagueService.Leagues);
+            TrayIcon.SendNotification("Sidekick is ready", $"Press {KeybindSetting.PriceCheck.GetTemplate()} over an item in-game to use. Press {KeybindSetting.CloseWindow.GetTemplate()} to close overlay.");
```
[the remark this answers is not in the document]
Comment (leMicin) [reply src/Sidekick/Services/TrayService.cs]: Fixed

File: src/Sidekick.Business/Apis/Poe/PoeApiService.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        }
+
+        public async Task<List<TReturn>> Fetch<TReturn>(string name, string path)
+            where TReturn : class
+        {
+            logger.Log($"Fetching {name} started.");
+            List<TReturn> result = null;
+            var success = false;
+
+            while (!success)
```
[the remark this answers is not in the document]
Comment (leMicin) [reply src/Sidekick.Business/Apis/Poe/PoeApiService.cs]: Is there anything to do here, or are you just happy with the solution?

File: src/Sidekick/Sidekick.csproj [hunk 3]
```diff
@@ -24,6 +24,7 @@
     
   <ItemGroup>
     <ProjectReference Include="..\Sidekick.Business\Sidekick.Business.csproj" />
+    <ProjectReference Include="..\Sidekick.Input\Sidekick.Keyboard.csproj" />
```
[the remark this answers is not in the document]
Comment (leMicin) [reply src/Sidekick/Sidekick.csproj]: Fixed

**Comment by Blinke** [comment]
> @Blinke
> I think your way wouldn't work. I think that when the DI tries to get the service for the "real" implementation would have a different instance than when you initialize with the other interfaces.

You're absolutely correct. That's what I get for thinking I have brilliant ideas when I have just gotten out of bed, haha.

**Comment by Blinke** [comment]
Actually, my early morning brain wasn't completely off track. The only change needed to make my solution work would be changing the line within the foreach to:

    services.AddSingleton(initializeInterface.GetType(), serviceProvider => serviceProvider.GetService<TInterface>());

That would get you the same instance for all the interfaces.

**Comment by leMicin** [comment]
@Blinke 
Oh I see, yeah that could work. You can make a PR with that change if you want. I will gladly accept that change :)

Though, please keep the TypeExtensions file, I will be using it in the future.

**Pull request merged**